Pyshow

cseed opened the pull request:
One issue to fix: doesn't escape string values correctly.  I will add a registry function for that.

The Scala version had a slight width miscalculation (started at fixedWidth instead of 2), otherwise character for character identical on a non-trivial VCF row table example.

tpoterba commented:
you are really cranking these out. I now feel useless.

akotlar commented on hail/python/hail/table.py:
```diff
@@ -1168,6 +1168,109 @@ def write(self, output: str, overwrite = False, stage_locally: bool = False,
+                start = i
+                w = column_width[i] + 4
+            i = i + 1
+        assert i == n_fields
```
Do we have any plans to throw pretty errors?

tpoterba replied:
this is a true assertion - there's nothing the user can do that should trigger it.

akotlar replied:
In that case, could it be placed in a unit test, to avoid runtime penalty (though small)?

tpoterba replied:
These kinds of assertions do ensure that our code doesn't enter a bad state, but also serve to provide some developer documentation. They're much better than written docs for enumerating invariants, because they're part of the code! Certainly worth the negligible performance hit.

They definitely don't take the place of unit tests, though.

tpoterba commented on hail/python/hail/table.py:
```diff
@@ -2143,7 +2249,8 @@ def _flatten(prefix, s):
                 return [(prefix, s)]
-        t = self.key_by()
+        # unkey but preserve order
+        t = self.order_by(*self.key)
```
wait, doesn't key_by() preserve order?

cseed replied:
No, I was also confused.  In particular, this rule:

```
    case TableKeyBy(TableKeyBy(child, _, _), keys, false) if canRepartition =>
      TableKeyBy(child, keys, false)
```

tpoterba replied:
Hmm....

I guess we don't make any guarantees about the relative order within identical keys.

But it does feel a bit weird that:

```
ht = ht.key_by('locus')
ht = ht.key_by()
```

will ignore the first key_by

cseed replied:
I think it makes sense now.  `t.key_by(k)` says make sure the result is keyed by `k`, but nothing more.  In particular, there is no guarantee of stability.

tpoterba replied:
yeah, I think it's fine.

I imagine this is a bug in other places, though.

tpoterba replied:
we will only guarantee order after an order_by, right?

cseed replied:
No, we guarantee the order for keyed things, too.  (Although we might stop that in 0.3 when we have a richer set of partitioners and keys should just be for join semantics.)

tpoterba commented:
failing doctests now.

tpoterba commented on hail/src/main/scala/is/hail/expr/ir/TableIR.scala:
```diff
@@ -1261,6 +1260,16 @@ case class TableOrderBy(child: TableIR, sortFields: IndexedSeq[SortField]) exten
+    if (sortFields.isEmpty)
+      return prev
+
+    if (sortFields.length <= prev.typ.key.length &&
```
actually, this should look at the physical key, to be even stronger.

cseed replied:
Oops, I had intended to do that, good catch.

tpoterba commented on hail/src/main/scala/is/hail/expr/ir/TableIR.scala:
```diff
@@ -1261,6 +1260,16 @@ case class TableOrderBy(child: TableIR, sortFields: IndexedSeq[SortField]) exten
   protected[ir] override def execute(hc: HailContext): TableValue = {
     val prev = child.execute(hc)
 
+    if (sortFields.isEmpty)
```
delete, handled by the case below.

cseed commented:
The final failures were doctests that called show on tables with duplicated keys, where the order of the duplicate rows, which we don't guarantee, were different.  I disabled those tests.  Also addressed comments, should be good to go.
